kroxylicious-filters/kroxylicious-encryption/doc/design.adoc

```diff
@@ -0,0 +1,267 @@
+
+=== KEK provisioning
+
+The encryption filter does not support provisioning its own KEKs using the KMS. Instead we assume that all KEKs are provisioned out of band.
+
+=== KEK selection and encryption policy
+
+KEK selection refers to how the filter determines which KEK to use for encrypting which records.
+The filter only supports this decision being based on the record (not, for example, based on the identity of the Kafka client), and without
+The filter only supports using a single key for encrypting record data.
+
+// === KEK rotation
+//
+// TODO
+//
+// === DEK generation
+//
+// TODO Talk about when and how new DEKs get generated
+//
+// === DEK rotation
+//
+// TODO
+
+
+== Encryption
+
+=== Supported ciphers
+
+Currently, the encryption filter supports
+
+* AES-GCM with 256 bit key, 96 bit initialization vector and 128 bit authentication tag.
+
+=== What to encrypt: The parcel
+
+We use the term _parcel_ to refer to the collection of data that gets encrypted.
+In other words, the parcel is the plaintext in the encryption operation.
+In version v1 we only support encrypting:
+
+* the record value
+* optionally, the record headers
+
+The following schema is used:
+[source,abnf]
+.The v1 parcel schema
+----
+parcel-v1              = record-value
+                         num-headers
+                         *header
+num-headers            = 1*OCTET ; unsigned VARINT
+header                 = name-length
+                         name
+                         value-length
+                         value
+name-length            = 1*OCTET ; unsigned VARINT
+name                   = *OCTET
+value-length           = 1*OCTET ; signed VARINT; -1 meaning null
+value                  = *OCTET
+record-value-length    = 1*OCTET ; signed VARINT; -1 meaning null
+record-value           = *OCTET
+----
+
+NOTE: `VARINT` refers to the same varint encoding that it used in the Kafka protocol. It is a variable length encoding of a 32 bit integer. Small integers are encoded as a single byte. In the worst case 5 byes are required for encoding.
+
+=== What to authenticate: The Additional Authenticated Data (AAD)
+
+Using AAD makes it harder (but not impossible) for a threat actor with write access to manipulate log segments without detection.
+
+NOTE: The Kafka record format prevents constructing an AAD schema that prevents all possible modifications to the log by a threat actor. For example, it is not practically possible to detect deletion of records from the log: For compacted logs this appears the same as legitimate record deletion by the log cleaner.
+
+There are currently two options for the use of Additional Authenticated Data (AAD):
+
+* None: No AAD
+* Batch: Associates some of the metadata of the batch within which the record is contained
++
+[source,abnf]
+.v1 AAD schema
+----
+aad-batch              = producer-id
```

Comment:
Note that I ended up calling this AAD, rather than AEAD as in previous documents. That's because AAD describes what this is. AEAD is a feature of some ciphers.